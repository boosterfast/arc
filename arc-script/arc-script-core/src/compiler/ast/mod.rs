#![allow(missing_docs)]

/// Module for debugging the AST.
pub(crate) mod debug;
/// Module for pretty printing the AST.
pub(crate) mod display;
/// Module which parses files of source code into an AST.
pub mod from;
/// Data representation of the AST.
pub mod repr;
pub(crate) mod utils;

/// Module for representing the AST data structure.
pub use repr::*;
