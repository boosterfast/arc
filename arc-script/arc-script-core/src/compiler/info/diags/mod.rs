pub mod repr;
pub mod to_codespan;
pub mod sink;
pub(crate) mod utils;

pub use repr::*;
pub use sink::*;
