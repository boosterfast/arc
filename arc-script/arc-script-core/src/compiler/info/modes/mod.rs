pub mod repr;
/// Module for representing sink and source information.
pub mod connector;

pub use repr::*;
