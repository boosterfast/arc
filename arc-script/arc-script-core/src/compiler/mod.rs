/// Module for representing Arc Queries.
pub(crate) mod query;
/// Module for representing the Abstract Syntax Tree.
///
/// NB: This module is public so that external libraries can construct ASTs
/// without having to generate source code.
pub mod ast;
/// Module for representing Dataflow graphs. Exports data types for staging
/// `HIR` functions.
pub mod dfg;
/// Module for representing Higher Order Intermediate Representation code.
pub(crate) mod hir;
/// Module for representing side information.
pub mod info;
/// Module for representing Multi-Level Intermediate Representation code.
pub(crate) mod mlir;
/// Module for representing Arcon code.
pub(crate) mod arcon;

// Module for incremental compilation.
// pub(crate) mod database;
/// Module which assembles compilation-pipeline.
pub mod pipeline;

pub(crate) mod pretty;

pub use pipeline::compile;
