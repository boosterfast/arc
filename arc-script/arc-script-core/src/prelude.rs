//! Exports for users who use the compiler as a library.

pub use crate::compiler;
pub use crate::compiler::ast;
pub use crate::compiler::info::diags;
pub use crate::compiler::info::logger;
pub use crate::compiler::info::modes;

pub use anyhow::Result;
pub use half::bf16;
pub use half::f16;
