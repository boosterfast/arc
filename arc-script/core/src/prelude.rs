pub use crate::ast::BinOpKind::*;
pub use crate::ast::DeclKind::*;
pub use crate::ast::DimKind::*;
pub use crate::ast::DimOpKind::*;
pub use crate::ast::ExprKind::*;
pub use crate::ast::LitKind::*;
pub use crate::ast::PatKind::*;
pub use crate::ast::ScalarKind::*;
pub use crate::ast::SettingKind::*;
pub use crate::ast::TypeKind::*;
pub use crate::ast::UnOpKind::*;
pub use crate::ast::*;
pub use crate::codegen::pretty::*;
pub use crate::error::CompilerError;
pub use crate::info::Info;
pub use crate::opt::*;
pub use crate::symbols::SymbolTable;
pub use crate::typer::Typer;
pub use codespan::Span;
pub use derive_more::Constructor;
