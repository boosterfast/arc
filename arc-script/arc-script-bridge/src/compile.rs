use arc_script_core::compiler::compile;
use arc_script_core::compiler::info::diags::sink::Buffer;
use arc_script_core::prelude::modes::{Input, Mode, Output};

use proc_macro as pm;
use quote::quote;

use std::fs;
use std::path::PathBuf;

/// arc_script::compile!("script.arc")
///    .stage_fun1(1, 2, 3)
///    .stage_fun2("foo", "bar")
///    .finalize()
pub(crate) fn expand(input: pm::TokenStream) -> pm::TokenStream {
    todo!()
}
