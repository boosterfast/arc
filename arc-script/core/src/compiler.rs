use crate::prelude::*;

pub fn diagnose(source: &str, opt: &Opt) {
    let script = compile(source, opt);
    if script.info.errors.is_empty() && !opt.debug {
        if opt.mlir {
            println!("{}", script.mlir());
        } else {
            println!("{}", script);
        };
    } else {
        script.emit_to_stdout()
    }
}

pub fn compile<'i>(source: &'i str, opt: &'i Opt) -> Script<'i> {
    let mut script = Script::parse(source, opt);

    if opt.debug {
        println!("=== Opt");
        println!("{:?}", opt);
        println!("=== Parsed");
        println!("{}", script);
    }

    // script.body.download();

    script.infer();

    if opt.debug {
        println!("=== Typed");
        println!("{}", script);
    }

    script = script.into_ssa();
    script.prune();

    if opt.debug {
        if opt.debug {
            println!("=== Canonicalized");
            println!("{}", script);
        }

        if script.info.errors.is_empty() {
            println!("=== MLIR");
            println!("{}", script.mlir());
        }
    }

    if !opt.check {
        let dataflow = script.eval();
        println!("{}", dataflow.pretty());
    }

    script
}
