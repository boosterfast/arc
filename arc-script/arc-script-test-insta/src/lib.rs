#![cfg_attr(not(test), allow(dead_code, unused_imports))]

use arc_script_core::prelude::compiler::compile;
use arc_script_core::prelude::diags::Buffer;
use arc_script_core::prelude::modes::{Input, Mode, Output};

use std::env;

fn snapshot(output: Output) {
    env::set_var("INSTA_OUTPUT", "summary");
    env::set_var("INSTA_FORCE_PASS", "1");
    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_snapshot_suffix();
    let mut sink = Buffer::no_color();
    settings.bind(|| {
        insta::glob!("{expect-pass, expect-fail, expect-fail-todo}/*", |path| {
            println!("Testing {}", path.display());
            let mode = Mode {
                input: Input::File(Some(path.into())),
                output,
                ..Default::default()
            };
            compile(mode, &mut sink).unwrap();
            let s = std::str::from_utf8(sink.as_slice()).unwrap();
            insta::assert_snapshot!(s);
            sink.clear();
        })
    });
}

#[test]
fn test_ast() {
    snapshot(Output::AST);
}

#[test]
fn test_hir() {
    snapshot(Output::HIR);
}

#[test]
fn test_dfg() {
    snapshot(Output::DFG);
}

#[test]
fn test_rust() {
    snapshot(Output::Rust);
}

#[test]
fn test_mlir() {
    snapshot(Output::MLIR);
}
