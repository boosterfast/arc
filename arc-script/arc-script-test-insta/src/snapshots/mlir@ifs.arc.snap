---
source: arc-script-test-insta/src/lib.rs
expression: s
input_file: arc-script-test-insta/src/expect-pass/ifs.arc
---
module @toplevel {
    func @main() -> i0 {
        %x_0 = arc.constant unit : i0
        return %x_0 : i0
    }
    func @test() -> si32 {
        %x_1 = arc.constant 65 : si32
        %x_2 = arc.constant 66 : si32
        %x_3 = constant true
        %x_4 = "arc.if"(%x_3) ({
            "arc.block.result"(%x_1) : (si32) -> ()
        },{
            "arc.block.result"(%x_2) : (si32) -> ()
        }) : (i1) -> si32
        return %x_4 : si32
    }
}

